proof-pile-v2/source_code/process_github.py

```diff
@@ -0,0 +1,331 @@
+        kws = {'pragma', 'endmodule', 'posedge', 'negedge', 'wire'}
+        for k in kws:
+            if k in example['text']:
+                return True
+        return False
+
+    def _has_bad_keyword(example):
+        kws = {'This file was automatically generated'}
+        for k in kws:
+            if k in example['text']:
+                return True
+        return False
+
+    keep = standard_filter(example)
+    keep = keep and _has_coq_keyword(example)
+    keep = keep and (not _has_verilog_keyword(example))
+    keep = keep and (not _has_bad_keyword(example))
+    return keep
+# ---
+
+
+def run(lang, file_pattern, filter_fn, limit, overwrite, dedup_chunk_size, out_dir):
+    print("Getting repos list...")
+    repos = get_repos(lang, limit, out_dir)
+    print("Downloading %d repos..." % (len(repos)))
+    for repo in tqdm(repos, total=len(repos)):
+        _get_dir_from_repo(**repo, overwrite=overwrite)
+        _delete_files_except_pattern(repo['save_path'], r".*\%s" % file_pattern)
+
+    _save_repo_metadata(repos, lang, out_dir)
+
+    print("Extracting files from repos...")
+    examples = []
+    for repo in tqdm(repos, total=len(repos)):
+        examples_ = _extract(repo['save_path'], "*%s" % file_pattern, metadata=repo)
+        examples.extend(examples_)
+    print("\t%d files" % len(examples))
+
+    print("Filtering...")
+    examples = _filter(examples, filter_fn)
+    print("\t%d files" % len(examples))
+    examples = deduplicate(examples, chunk_size=dedup_chunk_size)
+    print("\t%d files" % len(examples))
+
+    print("Computing tokens...")
+    num_tokens = token_length(examples)
+    print("\t%d tokens" % (num_tokens))
+
+    _save(examples, lang, out_dir, shard=0)  # extra copy before splitting; not strictly needed
+    _save_splits(
+        splits=make_splits(examples, args.eval_ratio),
+        out_dir=out_dir,
+        lang=lang
+    )
+    _save_stats({
+        'num_repos': args.limit,
+        'num_examples': len(examples),
+        'num_tokens': num_tokens,
+        'timestamp': datetime.now().strftime("%d/%m/%Y %H:%M:%S")
+    }, lang, out_dir)
+
+
+def coq(args):
+    run(
+        lang='coq',
+        file_pattern='.v',
+        filter_fn=filter_coq,
+        limit=args.limit,
+        overwrite=args.overwrite,
+        dedup_chunk_size=args.dedup_chunk_size,
+        out_dir=args.out_dir
+    )
+
+def setup(args):
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+
+
+def main(args):
+    if 'coq' in args.langs:
+        coq(args)
+
+
+if __name__ == '__main__':
+    import argparse
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--limit', type=int, default=250)
+    parser.add_argument('--langs', type=str, default=['coq'], nargs='+')
+    parser.add_argument('--dedup-chunk-size', type=int, default=2048)
+    parser.add_argument('--overwrite', action='store_true')
+    parser.add_argument('--out-dir', type=str, default='github-code')
```

Comment:
Maybe make the saving scheme consistent with `process_stack.py`? I.e data jsonls in `source_code/data_jsonl` and metadata in `source_code/meta_json`. 

Additionally, we should create a `coq_index` file that gets saved in `source_code/meta_json`, which is simply a file that contains the `{owner}/{name}` of a repository on each line (see `process_stack.py` for an example) . These get used when building the issues and diffs dataset.

Reply:
Will do (saving scheme and `coq_index`)!


Should we keep the `sha` somewhere?
Right now it saves a `repos-coq.jsonl` (I think `repo_dir` and `save_path` aren't necessary):
```
{"author": "AbsInt", "repo": "CompCert", "sha": "79ff6587290c207a8ba22adbe7bc4bfd5b478dde", "repo_dir": "", "save_path": "coq/AbsInt-CompCert"}
...
```